src/graft_dag.hpp

```diff
@@ -0,0 +1,178 @@
+  size_t GetEdgeIdx(const size_t parent_id, const size_t child_id) const;
+  // Get a sorted vector of all node subsplit's bitset representation. Optionally only
+  // graft nodes, or graft and host nodes.
+  BitsetVector GetSortedVectorOfNodeBitsets(bool include_host = true);
+  // Get a sorted vector of all edge PCSP's bitset representation. Optionally only graft
+  // edges, or graft and host edges.
+  BitsetVector GetSortedVectorOfEdgeBitsets(bool include_host = true);
+
+  // ** Counts
+
+  // Total number of taxa in DAG (same as Subsplit length).
+  size_t TaxonCount() const;
+  // Total number of nodes in full proposed DAG.
+  size_t NodeCount() const;
+  // Total number of nodes in graft only.
+  size_t GraftNodeCount() const;
+  // Total number of nodes in host DAG only.
+  size_t HostNodeCount() const;
+  // Total number of edges in full proposed DAG.
+  size_t EdgeCount() const;
+  // Total number of edges in graft only.
+  size_t GraftEdgeCount() const;
+  // Total number of edges in host DAG only.
+  size_t HostEdgeCount() const;
+  // Check if a node is from the host, otherwise from the graft.
+  bool IsNodeFromHost(size_t node_id) const;
+  // Check if an edge is from the host, otherwise from the graft.
+  bool IsEdgeFromHost(size_t edge_id) const;
+
+  // ** Contains
+
+  // Checks whether the node is in the host DAG or the graft.
+  bool ContainsNode(const Bitset node_subsplit) const;
+  bool ContainsNode(const size_t node_id) const;
+  // Checks whether the node is in the graft only.
+  bool ContainsGraftNode(const Bitset node_subsplit) const;
+  bool ContainsGraftNode(const size_t node_id) const;
+  // Checks whether the edge is in the host DAG or the graft.
+  bool ContainsEdge(const size_t parent_id, const size_t child_id) const;
+  // Checks whether the edge is in the graft only.
+  bool ContainsGraftEdge(const size_t parent_id, const size_t child_id) const;
+
+  // ** GP Operations
+  // These methods generate a serial vector of operations, but perform no computation.
+  // These operation vectors can then be computed and consumed by the
+  // GPEngine::ProcessOperations().
+
+  // Find all Quartet Hybrid Request object for given parent/child pair.  This finds the
+  // collection of rootward, sister, right and left quartet tips around focal edge.
+  QuartetHybridRequest QuartetHybridRequestOf(const size_t parent_id,
+                                              const bool is_focal_leftside,
+                                              const size_t child_id) const;
+
+  // ** Validity Tests
+
+  // Checks if host and graft are in a valid state.
+  // Specifically, checks that:
+  // - Host DAG is in valid state.
+  // - Each node Id corresponds to its array index.
+  // - Each node is valid. That is, either:
+  //    - Node has zero parents and zero children.
+  //    - Node has 1+ parents, 1+ sorted children, and 1+ rotated children.
```

Comment:
Old habits are hard to break! (sorted/rotated)

Reply:
Fixed.